samples/generate-offline-map-with-custom-parameters/src/main/java/com/esri/arcgismaps/sample/generateofflinemapwithcustomparameters/components/GenerateOfflineMapWithCustomParametersViewModel.kt

```diff
@@ -0,0 +1,386 @@
+                            if (isCropWaterPipesEnabled) {
+                                getGenerateGeodatabaseParameters(
+                                    parameterOverrides, getFeatureLayer("Main")
+                                )?.layerOptions?.forEach {
+                                    it.useGeometry = true
+                                }
+                            }
+                            // Get a reference to the hydrant layer
+                            getFeatureLayer("Hydrant")?.let { hydrantLayer ->
+                                // Get it's service layer id
+                                val serviceLayerId = getServiceLayerId(hydrantLayer)
+                                getGenerateGeodatabaseParameters(
+                                    parameterOverrides, getFeatureLayer(hydrantLayer.name)
+                                )?.layerOptions?.filter { it.layerId == serviceLayerId }?.forEach {
+                                    it.whereClause = "FLOW >= $minHydrantFlowRate"
+                                    it.queryOption = GenerateLayerQueryOption.UseFilter
+                                }
+                            }
+                            // Start a an offline map job from the task and parameters
+                            createOfflineMapJob(offlineMapTask, generateOfflineMapParameters, parameterOverrides)
+                        }
+
+                }
+        }
+    }
+
+    /**
+     * Generate an offline map job with the given [OfflineMapTask], [GenerateOfflineMapParameters] and
+     * [GenerateOfflineMapParameterOverrides].
+     */
+    private fun createOfflineMapJob(
+        offlineMapTask: OfflineMapTask,
+        generateOfflineMapParameters: GenerateOfflineMapParameters,
+        parameterOverrides: GenerateOfflineMapParameterOverrides
+    ) {
+        // Store the offline map in the app's scoped storage directory
+        val offlineMapPath = provisionPath + File.separator + "OfflineMap"
+        val offlineMapFile = File(offlineMapPath)
+
+        // Delete any offline map already present
+        offlineMapFile.deleteRecursively()
+        // Make the relevant directories for the offline map
+        offlineMapFile.mkdirs()
+
+        // Report any errors that occur during the offline map job
+        viewModelScope.launch(Dispatchers.Main) {
+            offlineMapTask.load().onFailure { error ->
+                messageDialogVM.showMessageDialog(
+                    title = error.message.toString(), description = error.cause.toString()
+                )
+            }
+        }
+
+        // Create an offline map job with the download directory path and parameters and start the job
+        generateOfflineMapJob = offlineMapTask.createGenerateOfflineMapJob(
+            parameters = generateOfflineMapParameters,
+            downloadDirectoryPath = offlineMapPath,
+            overrides = parameterOverrides
+        )
+
+        runOfflineMapJob()
```

Comment:
since you call `runOfflineMapJob` right after creating the `generateOfflineMapJob`, I would suggest passing this to `runOfflineMapJob` and displaying an useful message if the job fails for whatever reason. Then in `runOfflineMapJob` you won't have to use `?.`

Reply:
the cancel function needs access to the property level `generateOfflineMapJob`